samza-core/src/main/scala/org/apache/samza/diagnostics/DiagnosticsStreamMessage.java

```diff
@@ -0,0 +1,305 @@
+    MetricsSnapshot metricsSnapshot = new MetricsSnapshot(metricsHeader, new Metrics(metricsMessage));
+    return metricsSnapshot;
+  }
+
+  /**
+   * Check if the message has no contents.
+   * @return True if the message is empty, false otherwise.
+   */
+  public boolean isEmpty() {
+    return metricsMessage.isEmpty();
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    DiagnosticsStreamMessage that = (DiagnosticsStreamMessage) o;
+    return metricsHeader.getAsMap().equals(that.metricsHeader.getAsMap()) && metricsMessage.equals(that.metricsMessage);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(metricsHeader, metricsMessage);
+  }
+
+  public Collection<DiagnosticsManager.ProcessorStopEvent> getProcessorStopEvents() {
+    return (Collection<DiagnosticsManager.ProcessorStopEvent>) getFromMetricsMessage(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER,
+        STOP_EVENT_LIST_METRIC_NAME);
+  }
+
+  public Collection<DiagnosticsExceptionEvent> getExceptionEvents() {
+    return (Collection<DiagnosticsExceptionEvent>) getFromMetricsMessage(SAMZACONTAINER_METRICS_GROUP_NAME,
+        EXCEPTION_LIST_METRIC_NAME);
+  }
+
+  public int getContainerMb() {
+    return (int) getFromMetricsMessage(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER, CONTAINER_MB_METRIC_NAME);
+  }
+
+  public int getContainerNumCores() {
+    return (int) getFromMetricsMessage(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER, CONTAINER_NUM_CORES_METRIC_NAME);
+  }
+
+  public int getNumStoresWithChangelog() {
+    return (int) getFromMetricsMessage(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER,
+        CONTAINER_NUM_STORES_WITH_CHANGELOG_METRIC_NAME);
+  }
+
+  public Map<String, ContainerModel> getContainerModels() {
+    return deserializeContainerModelMap(
+        (Map<String, Object>) getFromMetricsMessage(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER, CONTAINER_MODELS_METRIC_NAME));
+  }
+
+  // Helper method to get a {@link DiagnosticsStreamMessage} from a {@link MetricsSnapshot}.
+  //   * This is typically used when deserializing messages from a diagnostics-stream.
+  //   * @param metricsSnapshot
+  public static DiagnosticsStreamMessage convertToDiagnosticsStreamMessage(MetricsSnapshot metricsSnapshot) {
+    DiagnosticsStreamMessage diagnosticsStreamMessage =
+        new DiagnosticsStreamMessage(metricsSnapshot.getHeader().getJobName(), metricsSnapshot.getHeader().getJobId(),
+            metricsSnapshot.getHeader().getContainerName(), metricsSnapshot.getHeader().getExecEnvironmentContainerId(),
+            metricsSnapshot.getHeader().getVersion(), metricsSnapshot.getHeader().getSamzaVersion(),
+            metricsSnapshot.getHeader().getHost(), metricsSnapshot.getHeader().getTime(),
+            metricsSnapshot.getHeader().getResetTime());
+
+    if (metricsSnapshot.getMetrics().getAsMap().containsKey(GROUP_NAME_FOR_DIAGNOSTICS_MANAGER)) {
```

Comment:
You could also get the GROUP_NAME_FOR_DIAGNOSTICS_MANAGER entry and reuse that. It might just be good to check if the GROUP_NAME_FOR_DIAGNOSTICS_MANAGER entry is null instead of checking `containsKey`, since a map can still "containsKey" and have a null value.